### PR TITLE
[rtcm3-rtk] RTK using ublox M8P-2

### PR DESCRIPTION
**Pull request by dewagter**
- [x] depends on https://github.com/paparazzi/pprzlink/pull/35

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -25,3 +25,6 @@
 [submodule "sw/ext/opencv_bebop"]
 	path = sw/ext/opencv_bebop
 	url = https://github.com/tudelft/opencv_bebop.git
+[submodule "sw/ext/librtcm3"]
```
Comment (gautierhattenberger) [inline .gitmodules]: why do we need a submodule for just one or two headers?

File: sw/airborne/modules/gps/gps_ubx_ucenter.c [hunk 2]
```diff
@@ -551,21 +562,45 @@ static bool gps_ubx_ucenter_configure(uint8_t nr)
       break;
     case 15:
       // Raw Measurement Data
-#if USE_GPS_UBX_RXM_RAW
```
Comment (gautierhattenberger) [inline sw/airborne/modules/gps/gps_ubx_ucenter.c]: better use `if` than `ifdef` here

File: sw/airborne/modules/gps/gps_ubx_ucenter.c [hunk 3]
```diff
@@ -551,21 +562,45 @@ static bool gps_ubx_ucenter_configure(uint8_t nr)
...
+#ifdef USE_GPS_UBX_RXM_RAW
       gps_ubx_ucenter_enable_msg(UBX_RXM_ID, UBX_RXM_RAW_ID, 1);
 #endif
       break;
     case 16:
       // Subframe Buffer
-#if USE_GPS_UBX_RXM_SFRB
```
Comment (gautierhattenberger) [inline sw/airborne/modules/gps/gps_ubx_ucenter.c]: better use `if` than `ifdef`

File: sw/airborne/modules/gps/gps_ubx_ucenter.c [hunk 4]
```diff
@@ -551,21 +562,45 @@ static bool gps_ubx_ucenter_configure(uint8_t nr)
...
+      gps_ubx_ucenter_enable_msg(UBX_RXM_ID, UBX_RXM_RTCM_ID, 1);
+      break;
+    case 21:
+#else
+    case 17:
+#endif
       // Try to save on non-ROM devices...
       UbxSend_CFG_CFG(gps_ubx_ucenter.dev, 0x00000000, 0xffffffff, 0x00000000);
       break;
+#ifdef USE_GPS_UBX_RTCM
```
Comment (gautierhattenberger) [inline sw/airborne/modules/gps/gps_ubx_ucenter.c]: better use `if` than`ifdef`

File: sw/airborne/subsystems/datalink/datalink.c [hunk 5]
```diff
@@ -127,8 +127,18 @@ void dl_parse_msg(struct link_device *dev, struct transport_tx *trans, uint8_t *
         );
       }
       break;
-#endif  // USE_GPS
-
+#ifdef USE_GPS_UBX_RTCM
```
Comment (gautierhattenberger) [inline sw/airborne/subsystems/datalink/datalink.c]: better use `if` than `ifdef`

File: sw/airborne/subsystems/gps.c [hunk 6]
```diff
@@ -282,6 +310,18 @@ void gps_init(void)
   register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_GPS_LLA, send_gps_lla);
   register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_GPS_SOL, send_gps_sol);
   register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_SVINFO, send_svinfo);
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_GPS_RTK, send_gps_rtk);
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_GPS_RXMRTCM, send_gps_rxmrtcm);
+#endif
+#ifdef USE_GPS_UBX_RTCM
```
Comment (gautierhattenberger) [inline sw/airborne/subsystems/gps.c]: better use `if` than `ifdef`

File: sw/airborne/subsystems/gps/gps_ubx.c [hunk 7]
```diff
@@ -23,6 +23,9 @@
 #include "subsystems/gps/gps_ubx.h"
 #include "subsystems/abi.h"
 #include "led.h"
+#ifdef USE_GPS_UBX_RTCM
```
Comment (gautierhattenberger) [inline sw/airborne/subsystems/gps/gps_ubx.c]: better use `if` than `ifdef`

File: sw/airborne/subsystems/gps/gps_ubx.c [hunk 8]
```diff
@@ -51,10 +56,18 @@
 
 struct GpsUbx gps_ubx;
 
+extern int gps_ubx_ucenter_get_status(void);
+
 #if USE_GPS_UBX_RXM_RAW
 struct GpsUbxRaw gps_ubx_raw;
 #endif
 
+#ifdef USE_GPS_UBX_RTCM
```
Comment (gautierhattenberger) [inline sw/airborne/subsystems/gps/gps_ubx.c]: better use `if` than `ifdef`

File: sw/airborne/subsystems/gps/gps_ubx.c [hunk 9]
```diff
@@ -157,13 +169,53 @@ void gps_ubx_read_message(void)
         gps_ubx.state.svinfos[i].azim = UBX_NAV_SVINFO_Azim(gps_ubx.msg_buf, i);
       }
     } else if (gps_ubx.msg_id == UBX_NAV_STATUS_ID) {
+#ifndef USE_GPS_UBX_RTCM
```
Comment (gautierhattenberger) [inline sw/airborne/subsystems/gps/gps_ubx.c]: better use `if` than `ifdef`

File: sw/airborne/subsystems/gps/gps_ubx.c [hunk 10]
```diff
@@ -180,8 +232,46 @@ void gps_ubx_read_message(void)
         gps_ubx_raw.measures[i].lli = UBX_RXM_RAW_lli(gps_ubx.msg_buf, i);
       }
     }
+#ifdef USE_GPS_UBX_RTCM
```
Comment (gautierhattenberger) [inline sw/airborne/subsystems/gps/gps_ubx.c]: something is missing here ?

File: sw/airborne/subsystems/gps/gps_ubx.c [hunk 11]
```diff
@@ -180,8 +232,46 @@ void gps_ubx_read_message(void)
...
+    else
+#endif // USE_GPS_UBX_RXM_RAW
+#endif // USE_GPS_UBX_RTCM
+#ifdef USE_GPS_UBX_RTCM
```
Comment (gautierhattenberger) [inline sw/airborne/subsystems/gps/gps_ubx.c]: better use `if` than `ifdef`

File: sw/ground_segment/misc/rtcm2ivy.c [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Paparazzi; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+/**
+ * \file rtcm2ivy.c
```
Comment (gautierhattenberger) [inline sw/ground_segment/misc/rtcm2ivy.c]: fix indentation for C files

File: sw/airborne/subsystems/gps/gps_ubx.c [hunk 13]
```diff
@@ -90,7 +102,7 @@ void gps_ubx_read_message(void)
       gps_ubx_time_sync.t0_tow_frac   = UBX_NAV_SOL_Frac(gps_ubx.msg_buf);
       gps_ubx.state.tow        = UBX_NAV_SOL_ITOW(gps_ubx.msg_buf);
       gps_ubx.state.week       = UBX_NAV_SOL_week(gps_ubx.msg_buf);
-      gps_ubx.state.fix        = UBX_NAV_SOL_GPSfix(gps_ubx.msg_buf);
+      //gps_ubx.state.fix        = UBX_NAV_SOL_GPSfix(gps_ubx.msg_buf);
```
Comment (flixr) [inline sw/airborne/subsystems/gps/gps_ubx.c]: why comment this here?

File: .gitmodules (hunk 1 above)
Comment (dewagter) [inline .gitmodules]: This is a fundamental question: do you want to allow re-use of code in other projects? Than separate libraries are the way to go. But submodules are probably not the way to go.

What about if I make a very simply package manager kind of app that can be used to copy-paste folders from a remote github account with a specific version, so we do store where the remote project is located and which version is the latest tested, but we do not need to import the entire project. For more complex subprojects like chibi-OS we can still use a submodule.

-e.g. librtcm: lib-only and not the documentation
-e.g. iridium-dialer: python script only and not the hardware

File: sw/airborne/subsystems/gps/gps_ubx.c (hunk 13 above)
Comment (dewagter) [inline sw/airborne/subsystems/gps/gps_ubx.c]: Fix can now grow to values 4 and 5. But this line was obviously not good. Fixed in new version.

File: sw/airborne/subsystems/gps/gps_ubx.c [hunk 14]
```diff
@@ -157,14 +172,50 @@ void gps_ubx_read_message(void)
         gps_ubx.state.svinfos[i].azim = UBX_NAV_SVINFO_Azim(gps_ubx.msg_buf, i);
       }
     } else if (gps_ubx.msg_id == UBX_NAV_STATUS_ID) {
+#if ! USE_GPS_UBX_RTCM
       gps_ubx.state.fix = UBX_NAV_STATUS_GPSfix(gps_ubx.msg_buf);
+#endif
       gps_ubx.status_flags = UBX_NAV_STATUS_Flags(gps_ubx.msg_buf);
       gps_ubx.sol_flags = UBX_NAV_SOL_Flags(gps_ubx.msg_buf);
+    } else if (gps_ubx.msg_id == UBX_NAV_RELPOSNED_ID) {
+      uint8_t version = UBX_NAV_RELPOSNED_VERSION(gps_ubx.msg_buf);
```
Comment (wv-tud) [inline sw/airborne/subsystems/gps/gps_ubx.c]: Should be `#if USE_GPS_UBX_RTCM` here, else `struct gps_relposned` is undeclared when `USE_GPS_UBX_RTCM` is undefined

File: sw/airborne/subsystems/gps/gps_ubx.c [hunk 15]
```diff
@@ -157,14 +172,50 @@ void gps_ubx_read_message(void)
...
+            }
+          } else {
+            gps_ubx.state.fix = 3; // 3D
+          }
+        } else {
+          gps_ubx.state.fix = 0;
+        }
+        printf("GNSS Fix OK: %i\tDGNSS: %i\tRTK: %i\trelPosValid: %i\trefStationId: %i\n", gps_relposned.gnssFixOK,
+               gps_relposned.diffSoln, gps_relposned.carrSoln, gps_relposned.relPosValid, gps_relposned.refStationId);
+      }
```
Comment (wv-tud) [inline sw/airborne/subsystems/gps/gps_ubx.c]: Should be matching `#endif  //USE_GPS_UBX_RTCM` here

File: sw/airborne/subsystems/gps/gps_ubx.c [hunk 16]
```diff
@@ -342,3 +432,52 @@ void gps_ubx_msg(void)
...
+  for (i = 0; i < n; i++) {
+    dev->put_byte(dev->periph, 0, buff[i]);
+  }
+  dev->send_message(dev->periph, 0);
+  return;
+}
+/**
+ * Override the default GPS packet injector to inject the data
+ */
+void gpss_inject_data(uint8_t packet_id, uint8_t length, uint8_t *data)
```
Comment (wv-tud) [inline sw/airborne/subsystems/gps/gps_ubx.c]: Yes this can be changed back to gps_inject_data.

**Comment by gautierhattenberger** [comment]
please fix conflicts with master

**Comment by dewagter** [comment]
- rebased & solved conflicts.
 - not test-flown with this code

**Comment by dewagter** [comment]
@gautierhattenberger are you also doing the reintegration of rtk: https://github.com/enacuavlab/paparazzi/tree/rtk-rtcm ?

File: sw/airborne/subsystems/datalink/datalink.c [hunk 17]
```diff
@@ -127,6 +127,17 @@ void dl_parse_msg(struct link_device *dev, struct transport_tx *trans, uint8_t *
         );
       }
       break;
+#if USE_GPS_UBX_RTCM
+      case DL_RTCM_INJECT : {
+              // GPS parse data
```
Comment (flixr) [inline sw/airborne/subsystems/datalink/datalink.c]: plz fix indentation

File: sw/airborne/subsystems/gps/gps_ubx.c [hunk 18]
```diff
@@ -24,6 +24,22 @@
...
+#ifndef USE_GPS_UBX_RTCM
+#define USE_GPS_UBX_RTCM 0
+#endif
+ 	
+
+#if USE_GPS_UBX_RTCM
+#include "subsystems/gps/librtcm3/CRC24Q.h"
+#endif
+
+#if PRINT_DEBUG_GPS_UBX_UCENTER
```
Comment (flixr) [inline sw/airborne/subsystems/gps/gps_ubx.c]: I would guess you want a different PRINT_DEBUG flag here...

**Comment by gautierhattenberger** [comment]
@dewagter my colleague needs to have this RTK stuff working for a project, but it is less advanced at least for the ground side (no decoding of rtcm frames). So I would say we keep your solution for master for now and if something is missing for us, we'll integrated on top.

**Comment by flixr** [comment]
@gautierhattenberger what is the status here?

**Comment by gautierhattenberger** [comment]
I don't know, I'll ask next week. But as I said, as soon as this one is good to merge you can do it. We'll do modification on top if needed.

**Comment by flixr** [comment]
@dewagter @gautierhattenberger this still fails because the `GPS_UBX_ENABLE_NMEA_DATA_MASK` define was removed... if that removal wasn't an accident, then it should be fixed in the code where it is still used...

**Comment by flixr** [comment]
ping @dewagter @gautierhattenberger

**Comment by gautierhattenberger** [comment]
work in progress....

**Comment by flixr** [comment]
So you will fix/update this PR?
Except for the now missing define I mentioned before it looks good to me...

**Comment by gautierhattenberger** [comment]
yes, I have someone testing it. It seems that he has some checksum issues.

**Comment by flixr** [comment]
Closed with #2034

**Pull request closed without merging**